### PR TITLE
feat: Add log service for embedded devices

### LINKED ISSUE #1: feat: Add page for showing device logging
The mangOH Red will fetch the device logging via diagnose tool and logread. The logging data should be shown on the data explorer.

### PR DESCRIPTION
**Pull request by splasky**
This commit implements the following API:

POST /register_device: Register embedded devices with IMEI and FSN.
The devices cannot insert logs if it is unregistered.

POST /grant_access_token: Get the json web token for insert log.

Close #1

### CODE REVIEW

File: config/config.go [hunk 1]
```diff
@@ -0,0 +1,5 @@
+package config
+
+const (
+	Domain = "localhost"
```
Comment (jserv) [inline config/config.go]: Can you read from configuration file?

File: controllers/controller.go [hunk 2]
```diff
@@ -41,6 +46,11 @@ type MAM_post_innder struct {
 	Chid string `json:"chid"`
 }
 
+type Device struct {
+	IMEI string `json:"imei"`
+	FSN  string `json:"fsn"`
```
Comment (jserv) [inline controllers/controller.go]: You should explain how to combine IMEI and FSN as an identifier.

File: templates/device_register_body.html [hunk 3]
```diff
@@ -0,0 +1,31 @@
+<div id="root">
```
Comment (jserv) [inline templates/device_register_body.html]: Can you shorten the HTML body? I am afraid of the depth.

**Comment by jserv** [comment]
If some commits were meant to be Work In Process, you should label `WIP`.

File: config.yml [hunk 4]
```diff
@@ -0,0 +1,4 @@
+DefaultSecret : ""
```
Comment (jserv) [inline config.yml]: Don't set prefix "default" since you already split the configuration file.

File: config.yml [hunk 5]
```diff
@@ -0,0 +1,5 @@
+SECRET: ""
+REDIS_URL: "localhost:6379"
```
Comment (jserv) [inline config.yml]: Change `REDIS_URL` to `REDIS_HOST`.

File: config.yml [hunk 6]
```diff
@@ -0,0 +1,5 @@
...
+REDIS_PASSWORD: ""
+REDIS_DB: 0
+ELASTICSEARCH_URL: "http://localhost:9200"
```
Comment (jserv) [inline config.yml]: Ditto. Change it to `ELASTICSEARCH_HOST`.

File: models/redis/redis.go [hunk 7]
```diff
@@ -0,0 +1,67 @@
...
+	"log"
+
+	"biilabs.com/mam-data-explorer/models/config"
+	"github.com/go-redis/redis"
+)
+
+func ConnectRedisDB() *redis.Client {
+	rdb := redis.NewClient(&redis.Options{
+		Addr:     config.ConfigHandler.GetString("REDIS_URL"),
+		Password: config.ConfigHandler.GetString("REDIS_PASSWORD"),
```
Comment (jserv) [inline models/redis/redis.go:13]: Would you like to store the information in certain file?

Comment (splasky) [reply models/redis/redis.go:13]: I think it is better to put all configuration together inside config.yml.

File: controllers/controller.go [hunk 8]
```diff
@@ -156,3 +161,178 @@ func Descrypt_mam_response(mam_message string, private_key string) Endpoint_obd2
 	log.Println("Descrypt mam response successfully")
 	return endpoint_data
 }
+
+func Register_device(c *gin.Context) {
+	var regDevice device.Device
+	if err := c.BindJSON(&regDevice); err != nil {
+		c.JSON(http.StatusOK, gin.H{
```
Comment (howjmay) [inline controllers/controller.go:167]: Should you add logger in error handling?

File: models/device/device.go [hunk 9]
```diff
@@ -0,0 +1,36 @@
...
+	err := json.Unmarshal([]byte(JSONDevice), &device)
+
+	if err != nil {
+		return nil, err
+
+	}
+	return &device, nil
+}
+
+// Concate IMEI and FSN as identity
```
Comment (howjmay) [inline models/device/device.go]: typo: concatenate

File: models/elasticsearch/elasticsearch.go [hunk 10]
```diff
@@ -0,0 +1,130 @@
...
+
+	type envelopeResponse struct {
+		Took int
+		Hits struct {
+			Total struct {
+				Value int
+			}
+			Hits []struct {
+				ID         string          `json:"_id"`
+				Source     json.RawMessage `json:"_source"`
```
Comment (howjmay) [inline models/elasticsearch/elasticsearch.go]: May I ask why do you need the underline?

**Review by howjmay — CHANGES_REQUESTED** [review]

**Comment by howjmay** [comment]
BTW this is such a huge change. Would you mind separating it into several small functional units?

File: controllers/controller.go (hunk 8 above)
Comment (splasky) [reply controllers/controller.go:167]: No, this is not the error. The user should post the correct json format.

File: models/elasticsearch/elasticsearch.go (hunk 10 above)
Comment (splasky) [reply models/elasticsearch/elasticsearch.go]: This is the json key defined by the elasticsearch response.
See also: https://github.com/elastic/go-elasticsearch

**Pull request merged**